components/public-api-server/pkg/auth/context.go

```diff
@@ -25,6 +25,8 @@ const (
 type Token struct {
 	Type  TokenType
 	Value string
+	// Only relevant for CookieTokenType
+	OriginHeader string
```

Comment:
IMO, the Token should not have the Origin defined. It's an unrelated piece of information. If we need, let's pull this out of the request in the connection constructor, but it doesn't need to be here.

Reply:
I'm also not super happy with this approach. But it's not unrelated: It's a required piece to decided whether we can trust the session token.

Reply:
I wondered if we can rename this struct? :thinking:

Reply:
Sure, but it's not an inherent property of the Token, it's a property of the request that we received. I should be able to use the token, without the Origin information and it's still a perfectly valid token.

Reply:
I'd rather see this be made part of the Server Proxy Conn constructor, as that's where it's required - when establishing the connection, not for the credentials passed into that connection.

Reply:
> I wondered if we can rename this struct? 🤔

What would that achieve? It's not intended to capture request information - for that we either use a request, or the context. It just abstracts over Cookie & Bearer credentials.

Reply:
That's fair. But having it as "required" parameter in the `NewCookieToken` method is a very nice property, and I'd like to keep that connectedness. It forces clients to think about the token. :slightly_smiling_face: 

To follow your idea: I thought about introduced a wrapping struct "Authentication" (no idea what the name would be?) that we place in Context instead, which contains the "Token" struct. But that introduced even more changes, and I was not sure the benefits are worth it.
Do you have another proposal that keeps the property of keeping a "connection between cookie and origin"?

Reply:
We can do that, but the Cookie, and the Origin header are not really connected. Yes, in browser scenarios you'd expect the Cookie to be sent only for the Origin, for which the Cookie was defined, but that does not mean I will not send a request which has a cookie and an entirely different origin (The Origin, and the Cookie behaviour are a client side behaviour, not a server side one.

If you want to make sure that connections to `server` over the protocol do send the origin, then this should be done in the Websocket connection constructor, ideally at the `gitpod-protocol` constructor level. This then forces you to supply it in the public-api implementation of the conn constructor, but that does not mean the token needs to encode it.

Reply:
@easyCZ Do you want to take over here? I have to focus on other things, and I feel we're mostly aligned.

I'd just love to not have to remember to get the token+origin header from somewhere and put it somewhere whenever we want to have a connection, but it should be somehow abstracted. E.g. `cookie, origin, err := auth.GetAuthenticationFromContext` _could_ work. Happy to give that a different name/separate structs/multi return/whatever :slightly_smiling_face:  :pray:

Reply:
If you think we have to change the token, then let's make the following change:

1. Token becomes an interface
```
type Token interface {
  Type() TokenType
  Value() string
  Origin() string
}
```
2. There are two implementations of this:
	* `CookieToken`, which has a helper constructor to do `NewCookieToken(cookie, origin)`
	* `BearerToken`, which has a helper constructor to do `NewBearerToken(token)`
3. Places that use the token can use the type to check, and call the interface methods to get the values.

This at least ensures that we're not growing the `Token` too arbitrarily

Reply:
> @easyCZ Do you want to take over here? I have to focus on other things, and I feel we're mostly aligned.

Sure, will push to this PR

Reply:
It does not have to be the "Token" struct. (discussing async is hard if we [write messages in parallel](https://github.com/gitpod-io/gitpod/pull/16405#discussion_r1106803971) :see_no_evil: )

Basically happy with whatever if we make it easy to do the right thing for clients. :lotus_position: